src/ledger/ledger.js

```diff
@@ -0,0 +1,344 @@
+    const updatedUser = {
+      id: existingUser.id,
+      name: existingUser.name,
+      aliases: updatedAliases,
+    };
+    this._users.set(userId, updatedUser);
+  }
+
+  /**
+   * Remove an alias from a user.
+   *
+   * Will no-op if the user doesn't have that alias.
+   *
+   * Will fail if the user does not exist.
+   *
+   * Will fail if the alias is in fact the user's innate address.
+   */
+  removeAlias(userId: UserId, alias: NodeAddressT): Ledger {
+    return this._act({type: "REMOVE_ALIAS", userId, alias});
+  }
+  _removeAlias(userId: UserId, alias: NodeAddressT) {
+    const existingUser = this._users.get(userId);
+    if (existingUser == null) {
+      throw new Error(`removeAlias: no user matching id ${userId}`);
+    }
+    if (alias === userAddress(userId)) {
+      throw new Error(`removeAlias: cannot remove user's innate address`);
+    }
+    const existingAliases = existingUser.aliases;
+    const idx = existingAliases.indexOf(alias);
+    if (idx === -1) {
+      // User is already not bound to this alias. No-op.
+      return;
+    }
+    const aliases = existingAliases.slice();
+    aliases.splice(idx, 1);
+    this._aliases.delete(alias);
+    this._users.set(userId, {
+      id: userId,
+      name: existingUser.name,
+      aliases,
+    });
+  }
+
+  /**
+   * Retrieve the log of all actions in the Ledger's history.
+   *
+   * May be used to reconstruct the Ledger after serialization.
+   */
+  actionLog(): LedgerLog {
+    return this._actionLog;
+  }
+
+  /**
+   * Reconstruct a Ledger from a LedgerLog.
+   */
+  static fromActionLog(log: LedgerLog): Ledger {
+    const ledger = new Ledger();
+    for (const a of log) {
+      ledger._act(a);
+    }
+    return ledger;
+  }
+
+  _act(a: Action): Ledger {
+    switch (a.type) {
+      case "ADD_USER":
+        this._addUser(a.user);
+        break;
+      case "RENAME_USER":
+        this._renameUser(a.userId, a.newName);
+        break;
+      case "REMOVE_USER":
+        this._removeUser(a.userId);
+        break;
+      case "ADD_ALIAS":
+        this._addAlias(a.userId, a.alias);
+        break;
+      case "REMOVE_ALIAS":
+        this._removeAlias(a.userId, a.alias);
+        break;
+      default:
+        throw new Error(`Unknown type: ${(a.type: empty)}`);
+    }
+    this._actionLog.push(a);
+    return this;
+  }
+}
+
+/**
+ * The log of all Actions in the Ledger.
+ *
+ * This is an opaque type; clients must not modify the log, since they
+ * could put it in an inconsistent state.
+ */
+export opaque type LedgerLog = $ReadOnlyArray<Action>;
+
+/**
+ * The Actions are used to store the history of Ledger changes.
+ */
+type Action = AddUser | RenameUser | RemoveUser | AddAlias | RemoveAlias;
+
+type AddUser = {|
+  +type: "ADD_USER",
+  +user: User,
+|};
+type RenameUser = {|
+  +type: "RENAME_USER",
+  +userId: UserId,
+  +newName: Username,
+|};
+type RemoveUser = {|
+  +type: "REMOVE_USER",
+  +userId: UserId,
+|};
```

Comment:
I think of it as giving people autonomy--contributors should have the right to delete their account. Although these might not be the right semantics, since once the account is deleted, their aliases (if any) will go back to collecting Grain independently. So perhaps it should be something that marks the user deleted and prevents them from showing up in results or earning more Grain, but keeps lets them keep hold of their addresses.

Reply:
I decided to remove `removeUser` for now. Semantics aren't clear enough, and we don't need it yet.